File: src/object.cpp
#include "object.h"

template <typename T> T& Object::getValue() {
    if (this->type == ObjectType::Pointer) {
        // Explicitly pass the template type T
        return this->object->getValue<T>();
    } else if constexpr (std::is_same_v<T, int64_t>) {
        if (this->type == ObjectType::Int) {
            return this->intData;
        }
    } else if constexpr (std::is_same_v<T, double>) {
        if (this->type == ObjectType::Double) {
            return this->doubleData;
        }
    } else if constexpr (std::is_same_v<T, std::string>) {
        if (this->type == ObjectType::String) {
            return this->stringData;
        }
    } else if constexpr (std::is_same_v<T, bool>) {
        if (this->type == ObjectType::Bool) {
            return this->boolData;
        }
    } else if constexpr (std::is_same_v<T, std::vector<Object>>) {
        if (this->type == ObjectType::List) {
            return this->listData;
        }
    } else if constexpr (std::is_same_v<T, std::map<Object, Object>>) {
        if (this->type == ObjectType::Map) {
            return this->mapData;
        }
    }

    throw std::bad_cast();
}

ObjectType Object::getType() const {
    if (this->type == ObjectType::Pointer) {
        return this->object->getType();
    }

    return this->type;
}

Object Object::toString() const {
    Object newObject = Object(this->toStr());
    return newObject; // Add this return statement
}

std::string Object::toStr() const {
    std::string stringValue;
    
    switch (this->type) {
        case ObjectType::Int:
            stringValue = std::to_string(this->intData);
            break;
        
        case ObjectType::Double:
            stringValue = std::to_string(this->doubleData);
            break;

        case ObjectType::Bool:
            if (this->boolData) {
                stringValue = "TRUE";
                break;
            }
            stringValue = "FALSE";
            break;

        case ObjectType::String:
            stringValue = this->stringData;
            break;

        case ObjectType::Void:
            stringValue = "Void";
            break;

        case ObjectType::List:
            stringValue = "[...]";
            break;
        
        case ObjectType::Map:
            stringValue = "{...}";
            break;

        case ObjectType::Pointer:
            // Correctly handle the pointer case
            if (this->object != nullptr) {
                stringValue = this->object->toStr();
                break;
            } else {
                stringValue = "NULL_POINTER";
                break;
            }

        default:
            stringValue = "|Unknown|";
            break;
    }
    
    return stringValue;
}

std::string Object::typeToStr() const {
    std::string typeName;
    
    switch (this->type) {
        case ObjectType::Int:
            typeName = "Int";
            break;
        
        case ObjectType::Double:
            typeName = "Double";
            break;

        case ObjectType::Bool:
            typeName = "Bool";
            break;

        case ObjectType::String:
            typeName = "String";
            break;

        case ObjectType::List:
            typeName = "List";
            break;
        
        case ObjectType::Map:
            typeName = "List";
            break;

        case ObjectType::Pointer:
            // Correctly handle the pointer case
            if (this->object != nullptr) {
                typeName = this->object->typeToStr();
                break;
            } else {
                typeName = "NULL_POINTER";
                break;
            }

        default:
            typeName = "|Unknown|";
            break;
    }

    if (this->nullable) {
        typeName += "?";
    }

    return typeName;
}

Object Object::multiply(Object& other) {
    if (!(other.getType() == ObjectType::Double || other.getType() == ObjectType::Int) || !(this->getType() == ObjectType::Double || this->getType() == ObjectType::Int)) {
        throw std::runtime_error("one or more values are non-numeric");
    }
    
    if (other.getType() == ObjectType::Double) {
        if (this->getType() == ObjectType::Double) {
            return Object(this->getValue<double>() * other.getValue<double>());
        }
        return Object(this->getValue<int64_t>() * other.getValue<double>());
    }

    if (this->getType() == ObjectType::Double) {
        return Object(this->getValue<double>() * other.getValue<int64_t>());
    }

    return Object(this->getValue<int64_t>() * other.getValue<int64_t>());
}

Object Object::add(Object& other) {
    if (!(other.getType() == ObjectType::Double || other.getType() == ObjectType::Int) || !(this->getType() == ObjectType::Double || this->getType() == ObjectType::Int)) {
        throw std::runtime_error("one or more values are non-numeric");
    }
    
    if (other.getType() == ObjectType::Double) {
        if (this->getType() == ObjectType::Double) {
            return Object(this->getValue<double>() + other.getValue<double>());
        }
        return Object(static_cast<double>(this->getValue<int64_t>()) + other.getValue<double>());
    }

    if (this->getType() == ObjectType::Double) {
        return Object(this->getValue<double>() + other.getValue<double>());
    }

    return Object(this->getValue<int64_t>() + other.getValue<int64_t>());
}

Object Object::divide(Object& other) {
    if (!(other.getType() == ObjectType::Double || other.getType() == ObjectType::Int) || !(this->getType() == ObjectType::Double || this->getType() == ObjectType::Int)) {
        throw std::runtime_error("one or more values are non-numeric");
    }

    if (other.getType() == ObjectType::Double) {
        if (this->getType() == ObjectType::Double) {
            return Object(this->getValue<double>() / other.getValue<double>());
        }
        return Object(static_cast<double>(this->getValue<int64_t>()) / other.getValue<double>());
    }

    if (this->getType() == ObjectType::Double) {
        return Object(this->getValue<double>() / other.getValue<double>());
    }

    return Object(static_cast<double>(this->getValue<int64_t>()) / other.getValue<double>());
}

Object Object::subtract(Object& other) {
    if (!(other.getType() == ObjectType::Double || other.getType() == ObjectType::Int) || !(this->getType() == ObjectType::Double || this->getType() == ObjectType::Int)) {
        throw std::runtime_error("one or more values are non-numeric");
    }
    
    if (other.getType() == ObjectType::Double) {
        if (this->getType() == ObjectType::Double) {
            return Object(this->getValue<double>() - other.getValue<double>());
        }
        return Object(static_cast<double>(this->getValue<int64_t>()) - other.getValue<double>());
    }

    if (this->getType() == ObjectType::Double) {
        return Object(this->getValue<double>() - other.getValue<double>());
    }

    return Object(this->getValue<int64_t>() - other.getValue<int64_t>());
}

Object Object::modulo(Object& other) {
    if (!(other.getType() == ObjectType::Int) || !(this->getType() == ObjectType::Int)) {
        throw std::runtime_error("one or more values are non-integer");
    }

    return Object(this->getValue<int64_t>() % other.getValue<int64_t>());
}

Object Object::intDivision(Object& other) {
    if (!(other.getType() == ObjectType::Int) || !(this->getType() == ObjectType::Int)) {
        throw std::runtime_error("one or more values are non-integer");
    }
    
    return Object(this->getValue<int64_t>() / other.getValue<int64_t>());
}

File: src/object.rs
use std::collections::HashMap;
use std::fmt;
use std::error::Error;

// Custom Error Type (Equivalent to throwing std::runtime_error/std::bad_cast)
#[derive(Debug)]
pub struct TypeError {
    pub message: String,
}

impl fmt::Display for TypeError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl Error for TypeError {}

// Corresponds to ObjectType enum
#[derive(Debug, Clone, PartialEq, Eq, Hash)] // Needs Hash for Map keys
pub enum ObjectData {
    Int(i64),
    Double(f64),
    String(String),
    Bool(bool),
    List(Vec<Object>),
    Map(HashMap<Object, Object>),
    // Note: Rust pointers are often handled via smart pointers or simply passing references
    // For this self-contained conversion, we'll omit a direct 'Pointer' variant for simplicity.
    Void,
}

// Corresponds to the Object class
#[derive(Debug, Clone, PartialEq, Eq, Hash)] // Needs Hash for Map keys
pub struct Object {
    // Note: Rust doesn't need C++'s `this->`
    pub data: ObjectData,
    pub constant: bool,
    pub nullable: bool,
}

impl Object {
    // --- Constructor-like functions ---

    pub fn new_int(value: i64, constant: bool, nullable: bool) -> Self {
        Object {
            data: ObjectData::Int(value),
            constant,
            nullable,
        }
    }
    
    // ... (other constructors like new_double, new_string, etc.)

    // Equivalent to Object::getType()
    pub fn get_type(&self) -> &str {
        match &self.data {
            ObjectData::Int(_) => "Int",
            ObjectData::Double(_) => "Double",
            ObjectData::String(_) => "String",
            ObjectData::Bool(_) => "Bool",
            ObjectData::List(_) => "List",
            ObjectData::Map(_) => "Map",
            ObjectData::Void => "Void",
        }
    }

    // Equivalent to Object::typeToStr()
    pub fn type_to_str(&self) -> String {
        let base_type = self.get_type().to_string();
        if self.nullable {
            format!("{}?", base_type)
        } else {
            base_type
        }
    }
    
    // --- Safe Value Access (Equivalent to Object::getValue<T>()) ---
    // Rust uses `Result` for error handling instead of exceptions.
    // This function returns a reference to the inner data if the types match.
    pub fn get_int(&self) -> Result<i64, TypeError> {
        if let ObjectData::Int(val) = self.data {
            Ok(val)
        } else {
            Err(TypeError { message: format!("Cannot cast {} to Int", self.get_type()) })
        }
    }
    
    // pub fn get_double(&self) -> Result<f64, TypeError> { ... }
    
    // --- Arithmetic Operations (Equivalent to Object::multiply, Object::add, etc.) ---
    
    pub fn multiply(&self, other: &Object) -> Result<Object, TypeError> {
        use ObjectData::*;
        
        match (&self.data, &other.data) {
            (Int(a), Int(b)) => Ok(Object::new_int(a * b, false, false)),
            (Int(a), Double(b)) => Ok(Object::new_double(*a as f64 * b, false, false)),
            (Double(a), Int(b)) => Ok(Object::new_double(a * *b as f64, false, false)),
            (Double(a), Double(b)) => Ok(Object::new_double(a * b, false, false)),
            _ => Err(TypeError { message: format!("Cannot multiply non-numeric types: {} and {}", self.get_type(), other.get_type()) }),
        }
    }

    // pub fn add(&self, other: &Object) -> Result<Object, TypeError> { ... }
    // pub fn divide(&self, other: &Object) -> Result<Object, TypeError> { ... }
    // pub fn subtract(&self, other: &Object) -> Result<Object, TypeError> { ... }

    pub fn modulo(&self, other: &Object) -> Result<Object, TypeError> {
        match (&self.data, &other.data) {
            (ObjectData::Int(a), ObjectData::Int(b)) => Ok(Object::new_int(a % b, false, false)),
            _ => Err(TypeError { message: "Modulo operation requires both operands to be Int".to_string() }),
        }
    }
}

// Trait implementation for printing (equivalent to toStr and toString)
impl fmt::Display for Object {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match &self.data {
            ObjectData::Int(val) => write!(f, "{}", val),
            ObjectData::Double(val) => write!(f, "{}", val),
            ObjectData::String(val) => write!(f, "{}", val),
            ObjectData::Bool(val) => write!(f, "{}", if *val { "TRUE" } else { "FALSE" }),
            ObjectData::Void => write!(f, "Void"),
            ObjectData::List(_) => write!(f, "[...]"),
            ObjectData::Map(_) => write!(f, "{{...}}"),
        }
    }
}